File: system_managerCPP/Src/RLPolicyClean.cpp
#include "RLPolicyClean.h"
#include "utils/SimpleJSONParser.h"
#include <iostream>
#include <stdexcept>

RLPolicyClean::RLPolicyClean() : gru_hidden_size(512), gru_input_size(0) {
    hxs = VectorXd::Zero(gru_hidden_size);
}

bool RLPolicyClean::load_normalizer_from_json() {
    if (!json_parser_) return false;
    
    std::string mean_key = "obs_normalizer.running_mean_std.running_mean_std.obs.running_mean";
    std::string var_key = "obs_normalizer.running_mean_std.running_mean_std.obs.running_var";
    
    std::vector<float> mean_data, var_data;
    std::vector<int> mean_shape, var_shape;
    
    if (json_parser_->extract_tensor(mean_key, mean_data, mean_shape) && 
        json_parser_->extract_tensor(var_key, var_data, var_shape)) {
        int obs_size = mean_shape.empty() ? mean_data.size() : mean_shape[0];
        obs_mean = VectorXd(obs_size);
        obs_var = VectorXd(obs_size);
        for (int i = 0; i < obs_size; i++) {
            obs_mean[i] = static_cast<double>(mean_data[i]);
            obs_var[i] = static_cast<double>(var_data[i]);
        }
        return true;
    }
    
    // Fallback: infer from encoder
    std::vector<int> indices;
    if (json_parser_->find_mlp_linear_indices(indices) && !indices.empty()) {
        std::string first_w_key = "encoder.encoders.obs.mlp_head." + std::to_string(indices[0]) + ".weight";
        std::vector<float> first_w_data;
        std::vector<int> first_w_shape;
        if (json_parser_->extract_tensor(first_w_key, first_w_data, first_w_shape) && first_w_shape.size() >= 2) {
            int obs_size = first_w_shape[1];
            obs_mean = VectorXd::Zero(obs_size);
            obs_var = VectorXd::Ones(obs_size);
            return true;
        }
    }
    
    // Last resort
    obs_mean = VectorXd::Zero(1);
    obs_var = VectorXd::Ones(1);
    return true;
}

bool RLPolicyClean::build_encoder_from_json(Activation nonlinearity) {
    if (!json_parser_) return false;
    
    encoder_layers_.clear();
    
    std::vector<int> indices;
    if (!json_parser_->find_mlp_linear_indices(indices)) {
        return true;  // No encoder layers - use identity
    }
    
    for (int idx : indices) {
        std::string w_key = "encoder.encoders.obs.mlp_head." + std::to_string(idx) + ".weight";
        std::string b_key = "encoder.encoders.obs.mlp_head." + std::to_string(idx) + ".bias";
        
        std::vector<float> w_data, b_data;
        std::vector<int> w_shape, b_shape;
        
        if (!json_parser_->extract_tensor(w_key, w_data, w_shape) || 
            !json_parser_->extract_tensor(b_key, b_data, b_shape)) {
            std::cerr << "Warning: Could not load encoder layer " << idx << std::endl;
            continue;
        }
        
        if (w_shape.size() != 2) {
            std::cerr << "Warning: Invalid weight shape for layer " << idx << std::endl;
            continue;
        }
        
        EncoderLayer layer;
        layer.linear.in_features = w_shape[1];
        layer.linear.out_features = w_shape[0];
        layer.activation = nonlinearity;
        
        // Reshape weight matrix: [out_features, in_features]
        layer.linear.weights.resize(w_shape[0]);
        for (int i = 0; i < w_shape[0]; i++) {
            layer.linear.weights[i].resize(w_shape[1]);
            for (int j = 0; j < w_shape[1]; j++) {
                layer.linear.weights[i][j] = w_data[i * w_shape[1] + j];
            }
        }
        
        layer.linear.bias = b_data;
        encoder_layers_.push_back(layer);
    }
    
    return true;
}

bool RLPolicyClean::load_gru_from_json() {
    if (!json_parser_) return false;
    
    std::string w_ih_key = "core.core.weight_ih_l0";
    std::string w_hh_key = "core.core.weight_hh_l0";
    std::string b_ih_key = "core.core.bias_ih_l0";
    std::string b_hh_key = "core.core.bias_hh_l0";
    
    std::vector<float> w_ih_data, w_hh_data, b_ih_data, b_hh_data;
    std::vector<int> w_ih_shape, w_hh_shape, b_ih_shape, b_hh_shape;
    
    if (!json_parser_->extract_tensor(w_ih_key, w_ih_data, w_ih_shape) ||
        !json_parser_->extract_tensor(w_hh_key, w_hh_data, w_hh_shape) ||
        !json_parser_->extract_tensor(b_ih_key, b_ih_data, b_ih_shape) ||
        !json_parser_->extract_tensor(b_hh_key, b_hh_data, b_hh_shape)) {
        return false;
    }
    
    gru_weights_.hidden_size = w_hh_shape[1];
    gru_weights_.input_size = w_ih_shape[1];
    
    // Reshape weight_ih: [3*hidden, input]
    int rows_ih = w_ih_shape[0];
    int cols_ih = w_ih_shape[1];
    gru_weights_.weight_ih.resize(rows_ih);
    for (int i = 0; i < rows_ih; i++) {
        gru_weights_.weight_ih[i].resize(cols_ih);
        for (int j = 0; j < cols_ih; j++) {
            gru_weights_.weight_ih[i][j] = w_ih_data[i * cols_ih + j];
        }
    }
    
    // Reshape weight_hh: [3*hidden, hidden]
    int rows_hh = w_hh_shape[0];
    int cols_hh = w_hh_shape[1];
    gru_weights_.weight_hh.resize(rows_hh);
    for (int i = 0; i < rows_hh; i++) {
        gru_weights_.weight_hh[i].resize(cols_hh);
        for (int j = 0; j < cols_hh; j++) {
            gru_weights_.weight_hh[i][j] = w_hh_data[i * cols_hh + j];
        }
    }
    
    gru_weights_.bias_ih = b_ih_data;
    gru_weights_.bias_hh = b_hh_data;
    gru_hidden_size = gru_weights_.hidden_size;
    
    // Resize hxs to match the loaded hidden size
    hxs = VectorXd::Zero(gru_hidden_size);
    
    // Determine input size from encoder
    if (!encoder_layers_.empty()) {
        gru_input_size = encoder_layers_.back().linear.out_features;
    } else {
        gru_input_size = gru_weights_.input_size;
    }
    
    return true;
}

bool RLPolicyClean::load_dist_linear_from_json() {
    if (!json_parser_) return false;
    
    std::string w_key = "action_parameterization.distribution_linear.weight";
    std::string b_key = "action_parameterization.distribution_linear.bias";
    
    std::vector<float> w_data, b_data;
    std::vector<int> w_shape, b_shape;
    
    if (!json_parser_->extract_tensor(w_key, w_data, w_shape) ||
        !json_parser_->extract_tensor(b_key, b_data, b_shape)) {
        return false;
    }
    
    dist_linear_.in_features = w_shape[1];
    dist_linear_.out_features = w_shape[0];
    
    // Reshape weight matrix: [out_features, in_features]
    dist_linear_.weights.resize(w_shape[0]);
    for (int i = 0; i < w_shape[0]; i++) {
        dist_linear_.weights[i].resize(w_shape[1]);
        for (int j = 0; j < w_shape[1]; j++) {
            dist_linear_.weights[i][j] = w_data[i * w_shape[1] + j];
        }
    }
    
    dist_linear_.bias = b_data;
    return true;
}

double RLPolicyClean::activation_func(double x, Activation act) const {
    switch (act) {
        case Activation::RELU:
            return std::max(0.0, x);
        case Activation::ELU:
            return x > 0.0 ? x : (std::exp(x) - 1.0);
        case Activation::TANH:
            return std::tanh(x);
    }
    return x;
}

VectorXd RLPolicyClean::linear_forward(const LinearLayer& layer, const VectorXd& input) const {
    VectorXd output(layer.out_features);
    
    for (int i = 0; i < layer.out_features; i++) {
        output[i] = layer.bias[i];
        for (int j = 0; j < layer.in_features; j++) {
            output[i] += layer.weights[i][j] * input[j];
        }
    }
    
    return output;
}

VectorXd RLPolicyClean::encoder_forward(const VectorXd& input) const {
    VectorXd x = input;
    
    for (const auto& layer : encoder_layers_) {
        x = linear_forward(layer.linear, x);
        for (int i = 0; i < x.size(); i++) {
            x[i] = activation_func(x[i], layer.activation);
        }
    }
    
    return x;
}

VectorXd RLPolicyClean::normalize_obs(const VectorXd& obs) const {
    VectorXd normalized(obs.size());
    for (int i = 0; i < obs.size(); i++) {
        double std_val = std::sqrt(obs_var[i] + EPS);
        normalized[i] = (obs[i] - obs_mean[i]) / std_val;
        normalized[i] = std::max(-5.0, std::min(5.0, normalized[i]));
    }
    return normalized;
}

VectorXd RLPolicyClean::gru_forward(const VectorXd& input) {
    int hidden = gru_hidden_size;
    
    // Split weights into z, r, n gates (each is hidden_size rows)
    auto get_gate_weights = [](const std::vector<std::vector<float>>& weights, int gate, int hidden_size) {
        std::vector<std::vector<float>> gate_weights(hidden_size);
        for (int i = 0; i < hidden_size; i++) {
            gate_weights[i] = weights[gate * hidden_size + i];
        }
        return gate_weights;
    };
    
    auto get_gate_bias = [](const std::vector<float>& bias, int gate, int hidden_size) {
        std::vector<float> gate_bias(hidden_size);
        for (int i = 0; i < hidden_size; i++) {
            gate_bias[i] = bias[gate * hidden_size + i];
        }
        return gate_bias;
    };
    
    // Matrix-vector multiply
    auto matvec = [](const std::vector<std::vector<float>>& W, const VectorXd& x) {
        VectorXd result(W.size());
        for (size_t i = 0; i < W.size(); i++) {
            result[i] = 0.0;
            for (int j = 0; j < x.size(); j++) {
                result[i] += W[i][j] * x[j];
            }
        }
        return result;
    };
    
    // Element-wise operations
    auto sigmoid = [](double x) { return 1.0 / (1.0 + std::exp(-x)); };
    auto tanh_func = [](double x) { return std::tanh(x); };
    
    // Get gate weights and biases
    // PyTorch GRU gate order is [reset, update, new]
    auto W_r = get_gate_weights(gru_weights_.weight_ih, 0, hidden);  // RESET gate
    auto W_z = get_gate_weights(gru_weights_.weight_ih, 1, hidden);  // UPDATE gate
    auto W_n = get_gate_weights(gru_weights_.weight_ih, 2, hidden);  // NEW gate
    
    auto U_r = get_gate_weights(gru_weights_.weight_hh, 0, hidden);  // RESET gate
    auto U_z = get_gate_weights(gru_weights_.weight_hh, 1, hidden);  // UPDATE gate
    auto U_n = get_gate_weights(gru_weights_.weight_hh, 2, hidden);  // NEW gate
    
    auto b_r_ih = get_gate_bias(gru_weights_.bias_ih, 0, hidden);  // RESET gate
    auto b_z_ih = get_gate_bias(gru_weights_.bias_ih, 1, hidden);  // UPDATE gate
    auto b_n_ih = get_gate_bias(gru_weights_.bias_ih, 2, hidden);  // NEW gate
    
    auto b_r_hh = get_gate_bias(gru_weights_.bias_hh, 0, hidden);  // RESET gate
    auto b_z_hh = get_gate_bias(gru_weights_.bias_hh, 1, hidden);  // UPDATE gate
    auto b_n_hh = get_gate_bias(gru_weights_.bias_hh, 2, hidden);  // NEW gate
    
    // Current hidden state
    VectorXd h_prev = hxs;
    
    // Compute gates
    VectorXd z_t = matvec(W_z, input);
    VectorXd r_t = matvec(W_r, input);
    VectorXd n_t = matvec(W_n, input);
    
    VectorXd z_h = matvec(U_z, h_prev);
    VectorXd r_h = matvec(U_r, h_prev);
    VectorXd n_h = matvec(U_n, h_prev);
    
    // Add biases and apply activations
    VectorXd z(hidden), r(hidden), n(hidden);
    for (int i = 0; i < hidden; i++) {
        z[i] = sigmoid(z_t[i] + z_h[i] + b_z_ih[i] + b_z_hh[i]);
        r[i] = sigmoid(r_t[i] + r_h[i] + b_r_ih[i] + b_r_hh[i]);
        // New gate: tanh(W_n @ x + b_n_ih + r * (U_n @ h + b_n_hh))
        n[i] = tanh_func(n_t[i] + b_n_ih[i] + r[i] * (n_h[i] + b_n_hh[i]));
    }
    
    // Update hidden state (PyTorch GRU formula)
    VectorXd h_next(hidden);
    for (int i = 0; i < hidden; i++) {
        h_next[i] = n[i] + z[i] * (h_prev[i] - n[i]);
    }
    
    hxs = h_next;
    return h_next;
}

std::pair<VectorXd, VectorXd> RLPolicyClean::forward(
    const VectorXd& obs,
    bool normalized) {
    
    // Save the input hidden state BEFORE it gets updated by gru_forward
    // This is the state that was used as input to produce the action
    VectorXd hxs_input = hxs;
    
    // Normalize observation
    VectorXd x = normalized ? obs : normalize_obs(obs);
    
    // Encoder forward
    x = encoder_forward(x);
    
    // GRU forward (this updates hxs to the output state)
    x = gru_forward(x);
    
    // Distribution linear
    VectorXd params = linear_forward(dist_linear_, x);
    
    // Split into mean and logstd, then concatenate
    int action_dim = params.size() / 2;
    VectorXd mean = params.head(action_dim);
    VectorXd logstd = params.tail(action_dim);
    
    VectorXd action_logits(params.size());
    action_logits.head(action_dim) = mean;
    action_logits.tail(action_dim) = logstd;
    
    // Return (action_logits, hxs_output) where hxs_output is the new state
    // But we also need to provide access to hxs_input for logging
    // For now, return the output state (hxs) as before
    return std::make_pair(action_logits, hxs);
}

void RLPolicyClean::reset_hidden_state(int batch_size) {
    (void)batch_size;  // Suppress unused parameter warning
    hxs = VectorXd::Zero(gru_hidden_size);
}

void RLPolicyClean::set_hidden_state(const VectorXd& hxs_vec) {
    if (hxs_vec.size() == gru_hidden_size) {
        hxs = hxs_vec;
    } else {
        std::cerr << "Warning: Hidden state size mismatch" << std::endl;
    }
}

std::shared_ptr<RLPolicyClean> RLPolicyClean::load_from_json(
    const std::string& json_path,
    const std::string& nonlinearity) {
    
    auto policy = std::make_shared<RLPolicyClean>();
    
    // Create and use SimpleJSONParser
    policy->json_parser_ = std::make_shared<SimpleJSONParser>();
    if (!policy->json_parser_->parse_file(json_path)) {
        throw std::runtime_error("Failed to parse JSON file: " + json_path);
    }
    
    // Convert nonlinearity string to enum
    Activation act = Activation::RELU;
    if (nonlinearity == "elu") {
        act = Activation::ELU;
    } else if (nonlinearity == "tanh") {
        act = Activation::TANH;
    }
    
    if (!policy->load_normalizer_from_json()) {
        std::cerr << "Warning: Could not load normalizer, using defaults" << std::endl;
    }
    
    if (!policy->build_encoder_from_json(act)) {
        throw std::runtime_error("Error: Could not build encoder");
    }
    
    if (!policy->load_gru_from_json()) {
        throw std::runtime_error("Error: Could not load GRU");
    }
    
    if (!policy->load_dist_linear_from_json()) {
        throw std::runtime_error("Error: Could not load distribution linear");
    }
    
    policy->reset_hidden_state(1);
    return policy;
}

File: system_managerCPP/python/rl_policy_clean.py
"""
Self-contained RL Policy implementation for inference testing.
Based on rl_policyClean.py but simplified for testing purposes.
"""
import sys
import os
import json
import base64

# Fix for Raspberry Pi Zero: disable SVE detection
os.environ['CPUINFO_DISABLE_SVE'] = '1'
import warnings
warnings.filterwarnings('ignore', category=RuntimeWarning, message='.*cpuinfo.*')
warnings.filterwarnings('ignore', message='.*prctl.*')

import torch
import torch.nn as nn
import numpy as np

EPS = 1e-5  # Match Sample Factory's _NORM_EPS


class RLPolicyClean(nn.Module):
    """Minimal, inference-only policy that mirrors Sample Factory forward pass.

    Components:
      - Input normalization using checkpoint running mean/var
      - Encoder MLP (Linear/ReLU × 3) for vector obs
      - GRU core
      - Linear distribution head producing [mean, logstd]

    Forward returns (action_logits, new_hxs) where action_logits = [mean, logstd].
    """

    def __init__(self):
        super().__init__()
        self.obs_mean: torch.Tensor | None = None
        self.obs_var: torch.Tensor | None = None
        self.encoder: nn.Module | None = None
        self.core: nn.GRU | None = None
        self.decoder: nn.Module | None = None
        self.dist_linear: nn.Linear | None = None
        self.hxs: torch.Tensor | None = None  # shape [num_layers, batch, hidden]

    def forward(self, obs: torch.Tensor, *, normalized: bool = False):
        """Forward pass through the policy network.
        
        Args:
            obs: Observation tensor [batch, features] or [features]
            normalized: If True, skip normalization (obs already normalized)
            
        Returns:
            action_logits: [batch, 2*action_dim] containing [mean, logstd] concatenated
            h_next: New hidden state [batch, hidden_size]
        """
        obs = torch.tensor(obs, dtype=torch.float32)
        if obs.dim() == 1:
            obs = obs.unsqueeze(0)  # [features] -> [1, features]
        
        # Normalize observation
        if not normalized:
            x = (obs - self.obs_mean) / torch.sqrt(self.obs_var + EPS)
            x = torch.clamp(x, -5.0, 5.0)
        else:
            x = obs

        # Encoder forward (standard PyTorch Sequential)
        x = self.encoder(x)  # [batch, encoder_out]

        # GRU forward (standard PyTorch GRU)
        # GRU expects [seq_len, batch, features]; hxs is [num_layers, batch, hidden]
        x = x.unsqueeze(0)  # [batch, features] -> [1, batch, features]
        if self.hxs is None or self.hxs.shape[1] != x.shape[1]:
            self.hxs = torch.zeros(1, x.shape[1], self.core.hidden_size, device=x.device)
        
        # Standard PyTorch GRU forward
        x, h_next = self.core(x, self.hxs)
        x = x.squeeze(0)  # [1, batch, hidden_size] -> [batch, hidden_size]
        self.hxs = h_next

        # Decoder forward (standard PyTorch module)
        if self.decoder is not None:
            x = self.decoder(x)

        # Distribution linear forward (standard PyTorch Linear)
        params = self.dist_linear(x)  # [batch, 2*action_dim]
        
        # Split into mean and logstd, then concatenate (standard PyTorch operations)
        mean, logstd = params.chunk(2, dim=-1)
        action_logits = torch.cat([mean, logstd], dim=-1)
        
        # Return h_next in [batch, hidden_size] format for consistency with C++
        h_next = h_next.squeeze(0)  # [1, batch, hidden_size] -> [batch, hidden_size]
        return action_logits, h_next

    def _init_modules(self, obs_mean: torch.Tensor, obs_var: torch.Tensor,
                       encoder: nn.Module, core: nn.GRU, dist_linear: nn.Linear,
                       decoder: nn.Module | None = None):
        self.obs_mean = nn.Parameter(obs_mean.float(), requires_grad=False)
        self.obs_var = nn.Parameter(obs_var.float(), requires_grad=False)
        self.encoder = encoder
        self.core = core
        self.decoder = decoder
        self.dist_linear = dist_linear
        self.hxs = None

    def reset_hidden_state(self, batch_size: int = 1, device: torch.device | None = None) -> None:
        """Reset internal RNN state to zeros."""
        if device is None:
            device = next(self.parameters()).device
        self.hxs = torch.zeros(1, batch_size, self.core.hidden_size, device=device)

    def set_hidden_state(self, hxs: torch.Tensor) -> None:
        """Set internal RNN state.
        
        Args:
            hxs: Hidden state tensor of shape [batch, hidden_size] or [1, batch, hidden_size]
        """
        if hxs.dim() == 2:
            self.hxs = hxs.unsqueeze(0).clone()
        else:
            self.hxs = hxs.clone()

    @staticmethod
    def _activation(nonlinearity: str) -> nn.Module:
        if nonlinearity == 'relu':
            return nn.ReLU(inplace=False)
        if nonlinearity == 'elu':
            return nn.ELU(inplace=False)
        if nonlinearity == 'tanh':
            return nn.Tanh()
        raise RuntimeError(f"Unsupported nonlinearity: {nonlinearity}")

    @staticmethod
    def _find_mlp_linear_indices(ckpt: dict) -> list[int]:
        """Find ordered Linear layer indices inside encoder.encoders.obs.mlp_head.*"""
        prefix = 'encoder.encoders.obs.mlp_head.'
        linear_indices: set[int] = set()
        for k in ckpt.keys():
            if not k.startswith(prefix):
                continue
            if k.endswith('.weight'):
                try:
                    idx = int(k[len(prefix):].split('.')[0])
                except Exception:
                    continue
                linear_indices.add(idx)
        return sorted(linear_indices)

    @classmethod
    def _build_encoder_from_ckpt(cls, ckpt: dict, *, nonlinearity: str, jit: bool) -> nn.Module:
        act = cls._activation(nonlinearity)
        indices = cls._find_mlp_linear_indices(ckpt)
        if len(indices) == 0:
            return nn.Identity()

        layers: list[nn.Module] = []
        for i, idx in enumerate(indices):
            w_key = f'encoder.encoders.obs.mlp_head.{idx}.weight'
            b_key = f'encoder.encoders.obs.mlp_head.{idx}.bias'
            weight = ckpt[w_key]
            bias = ckpt[b_key]
            in_f, out_f = weight.shape[1], weight.shape[0]
            lin = nn.Linear(in_f, out_f)
            lin.weight.data.copy_(weight)
            lin.bias.data.copy_(bias)
            layers.append(lin)
            layers.append(act)

        enc = nn.Sequential(*layers)
        if jit:
            enc = torch.jit.script(enc)
        return enc

    @classmethod
    def load_from_checkpoint(
        cls,
        path: str,
        device: str = 'cpu',
        *,
        nonlinearity: str = 'relu',
        jit_encoder: bool = False,
    ) -> 'RLPolicyClean':
        """Create a policy instance from a Sample Factory checkpoint (.pth).

        Args:
            path: checkpoint path (best_*.pth or checkpoint_*.pth)
            device: 'cpu' or 'cuda'
            nonlinearity: 'relu' | 'elu' | 'tanh' — must match training config
            jit_encoder: if True, torch.jit.script the encoder MLP
        """
        ckpt = torch.load(path, map_location=device, weights_only=False)["model"]

        # Observation normalizer parameters
        mean_key = 'obs_normalizer.running_mean_std.running_mean_std.obs.running_mean'
        var_key = 'obs_normalizer.running_mean_std.running_mean_std.obs.running_var'
        if mean_key in ckpt and var_key in ckpt:
            obs_mean = ckpt[mean_key].detach()
            obs_var = ckpt[var_key].detach()
        else:
            # Fallback: infer input size from the first linear layer
            indices = [0]
            try:
                indices = RLPolicyClean._find_mlp_linear_indices(ckpt)
            except Exception:
                pass
            if len(indices) > 0:
                first_w = ckpt[f'encoder.encoders.obs.mlp_head.{indices[0]}.weight']
                in_size = first_w.shape[1]
            else:
                in_size = 1
            obs_mean = torch.zeros(in_size)
            obs_var = torch.ones(in_size)

        # Build encoder
        encoder = cls._build_encoder_from_ckpt(ckpt, nonlinearity=nonlinearity, jit=jit_encoder)

        # Determine GRU input size
        gru_input_size = 512
        lin_indices = cls._find_mlp_linear_indices(ckpt)
        if len(lin_indices) > 0:
            last_idx = lin_indices[-1]
            last_w = ckpt[f'encoder.encoders.obs.mlp_head.{last_idx}.weight']
            gru_input_size = last_w.shape[0]

        # GRU core
        rnn_size = ckpt['core.core.weight_hh_l0'].shape[1]
        core = nn.GRU(input_size=gru_input_size, hidden_size=rnn_size, batch_first=False)
        core.weight_ih_l0.data.copy_(ckpt['core.core.weight_ih_l0'])
        core.weight_hh_l0.data.copy_(ckpt['core.core.weight_hh_l0'])
        core.bias_ih_l0.data.copy_(ckpt['core.core.bias_ih_l0'])
        core.bias_hh_l0.data.copy_(ckpt['core.core.bias_hh_l0'])

        # Identity decoder
        class _Identity(nn.Module):
            def forward(self, x):
                return x

        decoder = _Identity()

        # Distribution head: [mean, logstd]
        dist_in = ckpt['action_parameterization.distribution_linear.weight'].shape[1]
        num_of_actions = ckpt['action_parameterization.distribution_linear.weight'].shape[0]
        dist_linear = nn.Linear(dist_in, num_of_actions)
        dist_linear.weight.data.copy_(ckpt['action_parameterization.distribution_linear.weight'])
        dist_linear.bias.data.copy_(ckpt['action_parameterization.distribution_linear.bias'])

        policy = cls().to(device)
        policy._init_modules(obs_mean, obs_var, encoder, core, dist_linear, decoder)
        return policy

    @staticmethod
    def _decode_base64_tensor(tensor_dict: dict) -> torch.Tensor:
        """Decode tensor data from JSON format (supports both base64 and ASCII array formats).
        
        Args:
            tensor_dict: Dictionary with 'data' (base64 string or list of numbers), 'shape', 'dtype', 'encoding'
            
        Returns:
            PyTorch tensor
        """
        encoding = tensor_dict.get('encoding', None)
        data = tensor_dict['data']
        
        # Get dtype and determine numpy dtype
        dtype_str = tensor_dict.get('dtype', 'float32')
        if 'float64' in dtype_str or 'double' in dtype_str:
            np_dtype = np.float64
        elif 'float32' in dtype_str or 'float' in dtype_str:
            np_dtype = np.float32
        elif 'float16' in dtype_str or 'half' in dtype_str:
            np_dtype = np.float16
        else:
            np_dtype = np.float32  # default
        
        # Get shape
        shape = tuple(tensor_dict['shape'])
        
        # Handle base64 encoding
        if encoding == 'base64':
            # Decode base64 string
            decoded_bytes = base64.b64decode(data)
            # Convert bytes to numpy array
            num_elements = int(np.prod(shape))
            element_size = np.dtype(np_dtype).itemsize
            
            if len(decoded_bytes) != num_elements * element_size:
                raise ValueError(f"Size mismatch: expected {num_elements * element_size} bytes, got {len(decoded_bytes)}")
            
            np_array = np.frombuffer(decoded_bytes, dtype=np_dtype).reshape(shape)
        else:
            # Handle ASCII/array format - data is a list of numbers
            if isinstance(data, list):
                np_array = np.array(data, dtype=np_dtype)
            elif isinstance(data, str):
                # Try to parse as JSON array string
                import json
                data_list = json.loads(data)
                np_array = np.array(data_list, dtype=np_dtype)
            else:
                raise ValueError(f"Unsupported data format. Expected list or base64 string, got {type(data)}")
            
            # Reshape to match expected shape
            if np_array.shape != shape:
                np_array = np_array.reshape(shape)
        
        # Convert to PyTorch tensor
        return torch.from_numpy(np_array.copy())

    @classmethod
    def load_from_json(
        cls,
        json_path: str,
        device: str = 'cpu',
        *,
        nonlinearity: str = 'relu',
    ) -> 'RLPolicyClean':
        """Create a policy instance from a JSON file (created by pth2json.py).

        Args:
            json_path: Path to JSON file
            device: 'cpu' or 'cuda'
            nonlinearity: 'relu' | 'elu' | 'tanh' — must match training config
        """
        with open(json_path, 'r') as f:
            json_data = json.load(f)
        
        weights = json_data.get('weights', {})
        
        # Observation normalizer parameters
        mean_key = 'obs_normalizer.running_mean_std.running_mean_std.obs.running_mean'
        var_key = 'obs_normalizer.running_mean_std.running_mean_std.obs.running_var'
        
        if mean_key in weights and var_key in weights:
            obs_mean = cls._decode_base64_tensor(weights[mean_key])
            obs_var = cls._decode_base64_tensor(weights[var_key])
        else:
            # Fallback: infer input size from the first linear layer
            indices = cls._find_mlp_linear_indices_from_json(weights)
            if len(indices) > 0:
                first_w_key = f'encoder.encoders.obs.mlp_head.{indices[0]}.weight'
                first_w = cls._decode_base64_tensor(weights[first_w_key])
                in_size = first_w.shape[1]
            else:
                in_size = 1
            obs_mean = torch.zeros(in_size)
            obs_var = torch.ones(in_size)

        # Build encoder
        encoder = cls._build_encoder_from_json(weights, nonlinearity=nonlinearity)

        # Determine GRU input size
        gru_input_size = 512
        lin_indices = cls._find_mlp_linear_indices_from_json(weights)
        if len(lin_indices) > 0:
            last_idx = lin_indices[-1]
            last_w_key = f'encoder.encoders.obs.mlp_head.{last_idx}.weight'
            last_w = cls._decode_base64_tensor(weights[last_w_key])
            gru_input_size = last_w.shape[0]

        # GRU core
        rnn_size = cls._decode_base64_tensor(weights['core.core.weight_hh_l0']).shape[1]
        core = nn.GRU(input_size=gru_input_size, hidden_size=rnn_size, batch_first=False)
        core.weight_ih_l0.data.copy_(cls._decode_base64_tensor(weights['core.core.weight_ih_l0']))
        core.weight_hh_l0.data.copy_(cls._decode_base64_tensor(weights['core.core.weight_hh_l0']))
        core.bias_ih_l0.data.copy_(cls._decode_base64_tensor(weights['core.core.bias_ih_l0']))
        core.bias_hh_l0.data.copy_(cls._decode_base64_tensor(weights['core.core.bias_hh_l0']))

        # Identity decoder
        class _Identity(nn.Module):
            def forward(self, x):
                return x

        decoder = _Identity()

        # Distribution head: [mean, logstd]
        dist_w = cls._decode_base64_tensor(weights['action_parameterization.distribution_linear.weight'])
        dist_b = cls._decode_base64_tensor(weights['action_parameterization.distribution_linear.bias'])
        dist_in = dist_w.shape[1]
        num_of_actions = dist_w.shape[0]
        dist_linear = nn.Linear(dist_in, num_of_actions)
        dist_linear.weight.data.copy_(dist_w)
        dist_linear.bias.data.copy_(dist_b)

        policy = cls().to(device)
        policy._init_modules(obs_mean, obs_var, encoder, core, dist_linear, decoder)
        return policy

    @staticmethod
    def _find_mlp_linear_indices_from_json(weights: dict) -> list[int]:
        """Find ordered Linear layer indices from JSON weights dict."""
        prefix = 'encoder.encoders.obs.mlp_head.'
        linear_indices: set[int] = set()
        for k in weights.keys():
            if not k.startswith(prefix):
                continue
            if k.endswith('.weight'):
                try:
                    idx = int(k[len(prefix):].split('.')[0])
                except Exception:
                    continue
                linear_indices.add(idx)
        return sorted(linear_indices)

    @classmethod
    def _build_encoder_from_json(cls, weights: dict, *, nonlinearity: str) -> nn.Module:
        """Build encoder from JSON weights dict."""
        act = cls._activation(nonlinearity)
        indices = cls._find_mlp_linear_indices_from_json(weights)
        if len(indices) == 0:
            return nn.Identity()

        layers: list[nn.Module] = []
        for i, idx in enumerate(indices):
            w_key = f'encoder.encoders.obs.mlp_head.{idx}.weight'
            b_key = f'encoder.encoders.obs.mlp_head.{idx}.bias'
            weight = cls._decode_base64_tensor(weights[w_key])
            bias = cls._decode_base64_tensor(weights[b_key])
            in_f, out_f = weight.shape[1], weight.shape[0]
            lin = nn.Linear(in_f, out_f)
            lin.weight.data.copy_(weight)
            lin.bias.data.copy_(bias)
            layers.append(lin)
            layers.append(act)

        return nn.Sequential(*layers)
